asab/library/providers/filesystem.py

```
@@ -102,3 +117,71 @@ async def list(self, path: str) -> list:
 			))
 
 		return items
+
+
+	def _on_inotify_read(self):
+		data = os.read(self.FD, 64 * 1024)
+
+		pos = 0
+		while pos < len(data):
+			wd, mask, cookie, namesize = struct.unpack_from(EVENT_FMT, data, pos)
+			pos += EVENT_SIZE + namesize
+			name = (data[pos - namesize: pos].split(b'\x00', 1)[0]).decode()
+
+			if mask & IN_ISDIR == IN_ISDIR and ((mask & IN_CREATE == IN_CREATE) or (mask & IN_MOVED_TO == IN_MOVED_TO)):
+				subscribed_path, child_path = self.WDs[wd]
+				self._subscribe_recursive(subscribed_path, "/".join([child_path, name]))
+
+			if mask & IN_IGNORED == IN_IGNORED:
+				# cleanup
+				del self.WDs[wd]
+				continue
+
+			self.AggrEvents.append((wd, mask, cookie, os.fsdecode(name)))
+
+		self.AggrTimer.restart(0.2)
+
+
+	async def _on_aggr_timer(self):
+		to_advertise = set()
+		for wd, mask, cookie, name in self.AggrEvents:
+			# When wathed directory is being removed, more than one inotify events are being produced.
+			# When IN_IGNORED event occurs, respective wd is removed from self.WDs,
+			# but some other events (like IN_DELETE_SELF) get to this point, without having its reference in self.WDs.
+			subscribed_path, _ = self.WDs.get(wd, (None, None))
+			to_advertise.add(subscribed_path)
+		self.AggrEvents.clear()
+
+		for path in to_advertise:
+			if path is None:
+				continue
+			self.App.PubSub.publish("ASABLibrary.change!", self, path)
+
+
+	def subscribe(self, path):
+		if not os.path.isdir(self.BasePath + path):
```

subscribing to nonexisting directory (or file) should lead to silent error